flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,190 @@
+        waitAndAssert(
+                env,
+                buildStreamGraph(
+                        env,
+                        new ControlledSource() {
+                            @Override
+                            protected void beforeElement() throws Exception {
+                                if (getRuntimeContext().getAttemptNumber() == 0
+                                        && currentIndex == TOTAL_ELEMENTS / 2) {
+                                    waitWhile(() -> completedCheckpointNum.get() <= 0);
+                                    throwArtificialFailure();
+                                }
+                            }
+                        }));
+    }
+
+    /** Recovery from checkpoint containing non-materialized state and materialized state. */
+    @Test
+    public void testMaterialization() throws Exception {
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.enableCheckpointing(10)
+                .enableChangelogStateBackend(true)
+                .getCheckpointConfig()
+                .setCheckpointStorage(TEMPORARY_FOLDER.newFolder().toURI());
+
+        SerializableFunctionWithException<RunnableFuture<SnapshotResult<KeyedStateHandle>>>
+                snapshotResultConsumer =
+                        snapshotResultFuture -> {
+                            PENDING_MATERIALIZATION.add(snapshotResultFuture);
+                            return snapshotResultFuture;
+                        };
+        env.setStateBackend(
+                        new DelegatedStateBackendWrapper(
+                                delegatedStateBackend, snapshotResultConsumer))
+                .setRestartStrategy(RestartStrategies.fixedDelayRestart(2, 0));
+        env.configure(
+                new Configuration()
+                        .set(
+                                StateChangelogOptions.PERIODIC_MATERIALIZATION_INTERVAL,
+                                Duration.ofMillis(10)));
+        waitAndAssert(
+                env,
+                buildStreamGraph(
+                        env,
+                        new ControlledSource() {
+                            @Override
+                            protected void beforeElement() throws Exception {
+                                if (getRuntimeContext().getAttemptNumber() == 0
+                                        && currentIndex == TOTAL_ELEMENTS / 4) {
+                                    waitWhile(
+                                            () ->
+                                                    completedCheckpointNum.get() <= 0
+                                                            || PENDING_MATERIALIZATION.stream()
+                                                                    .noneMatch(Future::isDone));
+                                    PENDING_MATERIALIZATION.clear();
+                                    throwArtificialFailure();
```

Comment:
I think this is the same issue as [raised](https://github.com/apache/flink/pull/18224#discussion_r787022001) before:
> Completion of the Future doesn't guarantee that the backend is updated with the results. I guess the subsuquent sleep tries to address this (?); but they both use the Task thread, so it seems that backend doesn't (maybe I'm missing the goal or some thread details)

Reply:
Yes, you are right. I plan to add a condition to check the checkpoint status (like `ResumeCheckpointManuallyITCase#findExternalizedCheckpoint`), and the wrappered state backend will still be used  because I cannot find other way to make materialization fail(for `testFailedMaterialization`).
WDYT?

Reply:
I think this (checking that the checkpoint on the FS actually contains materialized state) should work.
And using a wrapper to fail materialization seems the right way to me.